rfcs/0006-configuration-constants.md

```diff
@@ -0,0 +1,125 @@
+- Feature Name: Configuration Time Constants, `config_time_constants`
+- Start Date: 2022-10-01
+- RFC PR: [FuelLabs/sway-rfcs#0006](https://github.com/FuelLabs/sway-rfcs/pull/19)
+- Sway Issue: [FueLabs/sway#1498](https://github.com/FuelLabs/sway/issues/1498)
+
+# Summary
+
+[summary]: #summary
+
+Configuration time constants can be conceptualized as traditional environment variables. Some bytecode has been compiled by the Sway compiler, and the SDK would like to configure some behavior of that bytecode with additional inputs that won't trigger a recompile.
+
+There is a similar feature that was implemented [here][pr_2549], but that was a mistaken interpretation of the requirements. In [#2549][pr_2549], a recompile still occurs and the new values are injected via `Forc.toml`, when we actually want these values to be injectable by the SDK.
+
+[pr_2549]: https://github.com/FuelLabs/sway/pull/2549
+
+# Motivation
+
+[motivation]: #motivation
+
+This supports the use case of configurable values post-compile-time but pre-runtime, similar to Solidity's `immutable` concept or environment variables in traditional programming. This allows for, e.g., updating contract addresses without recompiling the bytecode.
```

Comment:
It would be great to have some more fleshed out motivations here, otherwise it's tricky to weigh up the benefits vs the added complexity.

Currently, it looks like the only listed motivating example would already be addressed by https://github.com/FuelLabs/sway/pull/3016 in a manner that doesn't require a "configuration-time" step or new language features at all.

Reply:
I think the only motivations are avoiding recompiling from the SDK. @digorithm am I forgetting something?

Reply:
If our only motivations are avoiding recompilation, my next question would be: Why?

Is it just the compilation time that is bothersome? If so, I think it's worth noting that we haven't yet had a chance to assign any serious resources to bench-marking or optimizing compilation times beyond taking a quick look on the rare occasion they become unbearable. Perhaps it's worth considering investing the resources that would otherwise be involved in implementing/maintaining this new feature into optimizing compilation times?

I think your section on *Interactions with optimization passes* is an important one too, and I worry a little that we might end up accidentally blocking or over-complicating some future optimisation work unnecessarily.

In general, this feature seems big, complicated and annoying enough to maintain that it should probably have some strong motivations - I think we might want stronger motivations than we currently have?

@digorithm if you (or anyone else) have any more thorough motivating examples or workflows that would be great!